File: content/tracks/dsa/problems/lc-0271-encode-and-decode-strings/solution.py
from typing import List


class Codec:
    def encode(self, strs: List[str]) -> str:
        # Each string becomes "<length>#<string>": the length says where it ends, so the string
        # itself may contain '#' or digits.
        return "".join(f"{len(s)}#{s}" for s in strs)

    def decode(self, s: str) -> List[str]:
        result: List[str] = []
        i = 0
        while i < len(s):
            j = s.index("#", i)  # the first '#' from i closes the length prefix
            length = int(s[i:j])
            result.append(s[j + 1 : j + 1 + length])
            i = j + 1 + length
        return result

File: content/tracks/dsa/problems/lc-0271-encode-and-decode-strings/solution.go
package main

import (
	"strconv"
	"strings"
)

type Codec struct{}

func Constructor() Codec {
	return Codec{}
}

// Encode turns each string into "<length>#<string>": the length says where it ends, so the
// string itself may contain '#' or digits.
func (codec *Codec) Encode(strs []string) string {
	var encoded strings.Builder
	for _, str := range strs {
		encoded.WriteString(strconv.Itoa(len(str)))
		encoded.WriteByte('#')
		encoded.WriteString(str)
	}
	return encoded.String()
}

func (codec *Codec) Decode(s string) []string {
	result := []string{}
	for i := 0; i < len(s); {
		j := i + strings.IndexByte(s[i:], '#') // the first '#' from i closes the length prefix
		length, _ := strconv.Atoi(s[i:j])
		result = append(result, s[j+1:j+1+length])
		i = j + 1 + length
	}
	return result
}
